### ISSUE TITLE
New Lint: No recursion without tail calls

### ISSUE THREAD

**Issue by Evrey** [Manishearth/rust-clippy#1678]
Based on [this HIC++ guideline](http://www.codingstandard.com/rule/5-2-2-ensure-that-functions-do-not-call-themselves-either-directly-or-indirectly/).

The stack explodes much faster using recursion than a huge BFS work queue would explode the heap. Therefore, if clippy detects some direct or indirect recursion that does not benefit from tail call optimisation, clippy should issue a warning along with a short description about why this is potentially bad and a short hint on how to transform recursive algorithms to iterative ones *(e.g. using a work queue and BFS)*. And maybe a tl;dr about what tail calls are and how to make the algorithm tail recursive.

**Comment by llogiq** [comment]
I'd like to note that Rust does not have guaranteed tail call optimization (yet).

**Comment by Evrey** [comment]
I feared so. In that case the tail recursion hints would be a `TODO` for later.

**Comment by xfix** [comment]
And Rust cannot have _guaranteed_ tail call optimization. For instance, in this program, all function calls (except for `println!` macro) are tail recursive, and yet, they cannot be optimized, as I use stack space as sorta heap.

```rust
struct LinkedList<'a> {
    prev: Option<&'a LinkedList<'a>>,
    value: u32,
}

fn so_linked<F: Fn(LinkedList)>(linked: LinkedList, callback: F) {
    let value = linked.value;
    if value == 0 {
        callback(linked);
    } else {
        so_linked(
            LinkedList {
                prev: Some(&linked),
                value: value - 1,
            },
            callback,
        );
    }
}

fn main() {
    so_linked(
        LinkedList {
            prev: None,
            value: 99,
        },
        |ref list| {
            let mut list = list;
            loop {
                println!("{}", list.value);
                if let Some(prev) = list.prev {
                    list = prev;
                } else {
                    break;
                }
            }
        },
    )
}
```

**Comment by oli-obk** [comment]
@xfix guaranteed tail call optimization means that your code would fail to compile iff https://github.com/rust-lang/rfcs/pull/1888 get's accepted and you'd request tail call optimization by adding the `become` keyword to all tail call locations

**Comment by felix91gr** [comment]
@blyxyas this lint idea makes a lot of sense to me, but if I understand correctly Tail Call Optimizations are still pre-RFC: https://github.com/rust-lang/rfcs/pull/3407

Maybe this issue could be labeled as S-Blocked so that people know it isn't a possibility just yet, but it may be soon?

**Comment by blyxyas** [comment]
Although it's not guaranteed, if tail call optimization only happens *sometimes*, I think it's still worth-while to add as a lint for performance-critical code.

**Comment by felix91gr** [comment]
I think I agree. Though I don't think Clippy has the capability to see if a function will have its tail call eliminated by the backend (not until [RFC 3407](https://github.com/rust-lang/rfcs/pull/3407) gets merged at least).

So without that resource... 🤔 I guess it would be a direct recursion detection. Very similar to the lint I was working on before the feature freeze. Right?
